process/block/shardblock.go

```diff
@@ -1298,15 +1298,42 @@ func (sp *shardProcessor) updateState(headers []data.HeaderHandler, currentHeade
 			sp.accountsDB[state.UserAccountsState],
 		)
 
+		if sp.enableEpochsHandler.IsFlagEnabledInEpoch(common.EquivalentMessagesFlag, header.GetEpoch()) {
+			continue
+		}
+
 		sp.setFinalizedHeaderHashInIndexer(header.GetPrevHash())
 
-		finalRootHash := scheduledHeaderRootHash
-		if len(finalRootHash) == 0 {
-			finalRootHash = header.GetRootHash()
-		}
+		sp.setFinalBlockInfo(header, headerHash, scheduledHeaderRootHash)
+	}
 
-		sp.blockChain.SetFinalBlockInfo(header.GetNonce(), headerHash, finalRootHash)
+	if !sp.enableEpochsHandler.IsFlagEnabledInEpoch(common.EquivalentMessagesFlag, currentHeader.GetEpoch()) {
+		return
+	}
+
+	headerHash, err := core.CalculateHash(sp.marshalizer, sp.hasher, currentHeader)
+	if err != nil {
+		log.Debug("updateState.CalculateHash", "error", err.Error())
+		return
 	}
```

Comment:
pass curr headerHash as parameter to `updateState` ? instead of calculating it here again